src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,385 @@
+{
+    return m_parent->m_sections[m_index].key;
+}
+
+size_t ResultsSection::size()
+{
+    m_parent->calculate_sections_if_required();
+    REALM_ASSERT(m_parent->m_sections.size() > m_index);
+    auto range = m_parent->m_sections[m_index];
+    return range.indices.size();
+}
+
+NotificationToken ResultsSection::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                            KeyPathArray key_path_array) &
+{
+    return m_parent->add_notification_callback_for_section(m_index, std::move(callback), key_path_array);
+}
+
+realm::ThreadSafeReference ResultsSection::thread_safe_reference()
+{
+    return m_parent->thread_safe_reference();
+}
+
+bool ResultsSection::is_valid() const
+{
+    return m_parent->is_valid();
+}
+
+size_t ResultsSection::hash() const
+{
+    return m_parent->m_sections[m_index].hash;
+}
+
+SectionedResults::SectionedResults(Results results, SectionKeyFunc section_key_func)
+    : m_results(results)
+    , m_callback(std::move(section_key_func))
+{
+    calculate_sections_if_required(true);
+}
+
+SectionedResults::SectionedResults(Results results, Results::SectionedResultsOperator op,
+                                   util::Optional<StringData> prop_name)
+    : m_results(results)
+    , m_callback(builtin_comparison(results, op, prop_name))
+{
+    calculate_sections_if_required(true);
+}
+
+void SectionedResults::calculate_sections_if_required(bool force_update)
+{
+    if (m_results.m_update_policy == Results::UpdatePolicy::Never)
+        return;
+    else if (!m_results.has_changed() && !force_update)
+        return;
+
+    {
+        util::CheckedUniqueLock lock(m_results.m_mutex);
+        m_results.ensure_up_to_date();
+    }
+
+    calculate_sections();
+}
+
+// This method will run in the following scenarios:
+// - SectionedResults is freshly created from Results
+// - The underlying Table in the Results collection has changed
+void SectionedResults::calculate_sections()
+{
+    auto r = m_results.snapshot();
+    size_t size = r.size();
+    m_sections.clear();
+    m_row_to_index_path.clear();
+    std::map<Mixed, size_t> key_to_section_index;
+    m_row_to_index_path.resize(size);
+
+    for (size_t i = 0; i < size; ++i) {
+        Mixed key = m_callback(r.get_any(i), r.get_realm());
+        // Disallow links as section keys. It would be uncommon to use them to begin with
+        // and if the object acting as the key was deleted bad things would happen.
+        if (!key.is_null() && key.get_type() == type_Link) {
+            throw std::logic_error("Links are not supported as section keys.");
+        }
+        auto it = key_to_section_index.find(key);
+        if (it == key_to_section_index.end()) {
+            auto idx = m_sections.size();
+            Section section;
+            section.key = key;
+            section.index = idx;
+            section.indices.push_back(i);
+            section.hash = key.hash();
```

Comment:
This is still using a hash and need to not do that.

Reply:
What could be used in replacement for this? We would need to deep copy the value of the key otherwise.

Reply:
Hash has been removed, binary and string data keys will be stored in a buffer.